persist/fs/commitlog/read_write_prop_test.go

```diff
@@ -0,0 +1,369 @@
+
+package commitlog_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3db/context"
+	"github.com/m3db/m3db/persist/fs/commitlog"
+	"github.com/m3db/m3db/ts"
+	xtime "github.com/m3db/m3x/time"
+
+	"github.com/leanovate/gopter"
+	"github.com/leanovate/gopter/commands"
+	"github.com/leanovate/gopter/gen"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	defaultTestFlushInterval = time.Millisecond
+)
+
+func TestCommitLogReadWrite(t *testing.T) {
+	baseTestPath, err := ioutil.TempDir("", "commit-log-test-base-dir")
+	require.NoError(t, err)
+	defer os.RemoveAll(baseTestPath)
+
+	opts := commitlog.NewOptions()
+	fsOpts := opts.FilesystemOptions().SetFilePathPrefix(baseTestPath)
+	opts = opts.SetFilesystemOptions(fsOpts).SetFlushInterval(time.Millisecond)
+
+	cl := commitlog.NewCommitLog(opts)
+	require.NoError(t, cl.Open())
+
+	params := gopter.DefaultGenParameters()
+	writeResult := gen.SliceOfN(100, genWrite())(params)
+	writesInterface, ok := writeResult.Retrieve()
+	require.True(t, ok)
+	writes, ok := writesInterface.([]generatedWrite)
+
+	ctx := context.NewContext()
+	for _, w := range writes {
+		require.NoError(t, cl.WriteBehind(ctx, w.series, w.datapoint, w.unit, w.annotation))
+	}
+	ctx.Close()
+	require.NoError(t, cl.Close())
+
+	i := 0
+	iter, err := commitlog.NewIterator(opts)
+	require.NoError(t, err)
+	defer iter.Close()
+	for ; iter.Next(); i++ {
+		series, datapoint, _, _ := iter.Current()
+		write := writes[i]
+		require.Equal(t, write.series.ID.String(), series.ID.String())
+		require.Equal(t, write.series.Namespace.String(), series.Namespace.String())
+		require.Equal(t, write.series.Shard, series.Shard)
+		require.Equal(t, write.datapoint.Value, datapoint.Value)
+		require.True(t, write.datapoint.Timestamp.Equal(datapoint.Timestamp))
+	}
+	require.Equal(t, len(writes), i)
+}
+
+func TestCommitLogPropTest(t *testing.T) {
+	basePath, err := ioutil.TempDir("", "commit-log-tests")
+	require.NoError(t, err)
+	defer os.RemoveAll(basePath)
+
+	parameters := gopter.DefaultTestParameters()
+	parameters.MinSuccessfulTests = 30
+	properties := gopter.NewProperties(parameters)
+
+	comms := clCommandFunctor(basePath, t)
+	properties.Property("CommitLog System", commands.Prop(comms))
+	properties.TestingRun(t)
+}
+
+// clCommandFunctor is a var which implements the command.Commands interface,
+// i.e. is responsible for creating/destroying the system under test and generating
+// commands and initial states (clState)
+func clCommandFunctor(basePath string, t *testing.T) *commands.ProtoCommands {
+	return &commands.ProtoCommands{
+		NewSystemUnderTestFunc: func(initialState commands.State) commands.SystemUnderTest {
+			return initialState
+		},
+		DestroySystemUnderTestFunc: func(sut commands.SystemUnderTest) {
+			state := sut.(*clState)
+			if cl := state.cLog; cl != nil {
+				cl.Close()
+			}
+			os.RemoveAll(state.opts.FilesystemOptions().FilePathPrefix())
```

Comment:
Yep so need this one because gopter can re-create the same system multiple times, want each run to have a clean directory to work with. 

The other one is me being paranoid, in case there's a gopter failure mode we haven't run into yet.